### DISCUSSION TITLE
Is it a good idea to remove prototype from OM core?

### DISCUSSION THREAD

**Discussion by kiatng** [OpenMage/magento-lts#3645] [Ideas]
Jenkins announced in [May](https://www.jenkins.io/blog/2023/05/12/removing-prototype-from-jenkins/) that they are removing prototype.js, and they did it [last month](https://www.jenkins.io/blog/2023/10/09/prototype-removed/). Is this something we want to do? How to go about it?

**Comment by Flyingmana** [comment]
it will be a big backwards compatibility break and is likely only possible by building fully new Frontend.

but looking at the (in)activity https://github.com/prototypejs/prototype/graphs/contributors
and also the activity of the most active fork https://github.com/wiz78/prototype

I think it will be a requirement to switch away in the longterm.

**Comment by fballiano** [comment]
it's gonna be extremely hard but, I've said it in the past, IMHO it would be beautiful to have it removed.
pagespeed marks legacy libraries as a penalty and having something as old as prototypejs makes openmage seems still old.

**Comment by addison74** [comment]
I would like to get rid of the prototype and replace it with vanilla JS, jQuery or another maintained and popular library with plugins. If in the case of OpenMage with tailor-made work something like this can be achieved, in the case of extensions that use Prototype it will be a big problem. On the one hand, the developer will not want to update the extension/module, on the other hand, as soon as a store it is updated without Prototype, it will stop working on certain parts. Along with this important change, which I think is much more complex than the transition to ZF1-Future, we will also have to be offered a change guide, the new equivalents to what exists now in the extensions. nor can it be compared with the removal of MySQL4 parts from the code and extensions, which should be completed by the end and we can remove relics from 10 years ago. At this moment I am using 4 extensions that use Prototype.

If the community wants this, discussions can be initiated on what to replace it with, manual changes, . As in OpenMage we have jQuery and Magento 2 uses it, it can be a solution.

**Reply by luigifab** [reply]
[Redmine did it](https://www.redmine.org/issues/11445?tab=changesets) without questions!
But we can allow OpenMage to work without Prototypejs (front and back), and load prototype files with a configuration for compatibility.

**Comment by kiatng** [comment]
My idea is to replace it with vanilla js. As we don't have the resources to do it in one big ban, can we can do it piece by piece, file by file, similar to how we do PHP8? When all is done, we can still keep prototype in the layout head for 3rd-party extensions. For those who has no trace of prototype, they can easily remove it in custom layout file. For change guide, we can do similar like [Jenkins](https://www.jenkins.io/blog/2023/05/12/removing-prototype-from-jenkins/):
![image](https://github.com/OpenMage/magento-lts/assets/1106470/1513483e-1de6-458f-87e1-ab0590f41aad)

**Reply by Flyingmana** [reply]
Changing away from prototype is going to break a lot of 3th party code. Even if we let the library in, the core code we will change is enough to break, especially in the checkout area.

The most user-friendly way to do it would be by having a separate Theme, and still continuing to offer the old theme containing prototype for the next years.

**Reply by kiatng** [reply]
A separate theme adds to maintenance costs. The checkout uses a few js located in the skin. I am not sure if it's possible to replace prototype to vanilla in these js files.; but if there is a way to do it, would it break anything in `base` and `rwd`?

**Comment by kiatng** [comment]
I gather that we need to remove prototype.js. Now, how to do it?

1. Replace with vanilla js.
2. Do it file by file. PR should be small, should not have too many files. New PR tag: `deprecate prototype.js`
3. Do it in `main` branch.

What do you think?

**Reply by fballiano** [reply]
agreed 100%, on `next` we could also remove the script from the head block
